### ISSUE TITLE
@types/mapbox-gl: accessToken read-only?

### ISSUE THREAD

**Issue by dccarmo** [DefinitelyTyped/DefinitelyTyped#23467]
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @dobrud, @patrickr

Shouldn't `accessToken` writeable? Am I missing something?

**Comment by salmau** [comment]
You can use (mapboxgl as any).accessToken

**Comment by Doug-Shannon** [comment]
That works, but is the spec wrong here?  Seems like you should be able to write to it.

**Comment by ksylvest** [comment]
Yeah - this is definitely a bug. Annoying to have invalid type definition for the configuration (everyone needs to use this before using the library).

**Comment by rschulman** [comment]
I am also having this issue, and neither of the `(mapboxgl as any)` nor the `Object.getOwnPropertyDescriptor` answers from above seem to be working for me. Typescript isn't complaining, but the browser console is telling me that an API access token is required.

Anyone have any thoughts on what I could try next?

**Comment by garrettgsb** [comment]
Same problem. Functional workarounds for me are `(mapboxgl as any).accessToken = 'abc123'` and changing `import * as mapboxgl from 'mapbox-gl'` to `const mapboxgl = require('mapbox-gl')`.

The examples from Mapbox's own repos indicate that the `accessToken` property should indeed be writable. Example:

https://github.com/mapbox/mapbox-react-examples/blob/master/react-tooltip/src/index.js

**Comment by HeleleF** [comment]
Its been over a year now... will this be fixed?

**Comment by mgcarre** [comment]
Having the same issue accessToken is readonly...

Had this fixed modifying the declaration file

at line 13 instead of :
```javascript
export = mapboxgl;
```
I have :
```javascript
export default mapboxgl;
```

Then I imported the whole namespace which does not cause any trouble... Working fine for v.1.2.1
```javascript
import mapboxgl from 'mapbox-gl';
```

**Comment by stanleenwosu** [comment]
Worked for me

**Comment by herr0092** [comment]
+1 Same error here

**Comment by tomekr** [comment]
@mgcarre is that something that could get PR'd?

Another alternative which doesn't require any weirdness is to use the Map object's access token option:

```
import { Map } from 'mapbox-gl';

var map = new Map({
  accessToken: 'pk...',
  container: 'map',
  style: 'mapbox://styles/mapbox/streets-v11'
});

**Comment by davidgilbertson** [comment]
@tomekr I came here to say exactly this :)

It's odd that Mapbox use the `mapboxGl.accessToken = '...'`  pattern in all their examples when passing `accessToken` as an option is more straightforward.

**Comment by mgcarre** [comment]
@tomekr 👍 
I re-read the typescript documentation and found some interesting stuff...

First the doc says : "ES6 modules cannot directly export class objects [...] use the CommonJS-style:
``` typescript
import mapboxgl = require('mapbox-gl');
mapboxgl.accessToken = 'YOUR_TOKEN';
```

Instead, you can use the ES6 syntax **only** if you add the _esModuleInterop_ (affects code emit) or _allowSyntheticDefaultImports_ (no affects to code emit just type checking) to your typescript config file or flags... Example config file :
``` json
{
    "compilerOptions": {
        "esModuleInterop": true
    }
}
```
This automatically adds the "export default ..." to the proper file and allows you to use your mapboxgl object as before :
``` typescript
import mapboxgl from 'mapbox-gl';

mapboxgl.accessToken = 'YOUR_TOKEN';
```

In conclusion no PR is needed but right configuration file/flag 😄 

> Source : https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require

**Comment by dhivahari** [comment]
> @mgcarre is that something that could get PR'd?
> 
> Another alternative which doesn't require any weirdness is to use the Map object's access token option:
> 
> ```
> import { Map } from 'mapbox-gl';
> 
> var map = new Map({
>   accessToken: 'pk...',
>   container: 'map',
>   style: 'mapbox://styles/mapbox/streets-v11'
> });
> ```

This effin worked for me! I'm here at 3.40 AM with very little Angular experience, and this was very nice. ty!

**Comment by hmendezm** [comment]
Hi everybody,

I have mapbox-gl version 2.14.1 and I am getting this read-only problem in the property accessToken. For now, I am using (mapboxgl as any).accessToken = '??????'. is there any better solution to this?

Best
Hmendezm

**Comment by tangerren** [comment]
this type definition is very ugly

**Comment by sdn2525** [comment]
You can set this on the `ACCESS_TOKEN` property present on `mapboxgl.config`, where you can avoid both errors.
 
1. **Imports are Immutable** error during the compilation phase when directly setting this as 

    `mapboxgl.accessToken = 'YOUR_TOKEN';`

 2. **set is not a function** error on the browser when you have hosted your compiled artifact with the static method

    `Object. getOwnPropertyDescriptor(mapboxgl, "accessToken"). set(YOUR_TOKEN');`  

```
import * as mapboxgl from 'mapbox-gl';

mapboxgl.config.ACCESS_TOKEN = 'YOUR_TOKEN';
```

**Comment by sdn2525** [comment]
You can set this on the `ACCESS_TOKEN` property present on `mapboxgl.config`, where you can avoid both errors.
 
1. **Imports are Immutable** error during the compilation phase when directly setting this as 

    `mapboxgl.accessToken = 'YOUR_TOKEN';`

 2. **set is not a function** error on the browser when you have hosted your compiled artifact with the static method

    `Object.getOwnPropertyDescriptor(mapboxgl, "accessToken"). set(YOUR_TOKEN');`  

```
import * as mapboxgl from 'mapbox-gl';

mapboxgl.config.ACCESS_TOKEN = 'YOUR_TOKEN';
```
